### PR TITLE
A couple performance tweaks for animated images #trivial

### PR DESCRIPTION
**Pull request by garrettmoon**

### CODE REVIEW

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Any source code changes should have an entry in CHANGELOG.md or have #trivial in their title.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by pinterest-ios** [comment]
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/328e4aa8f986b768b2bc9e70e5e3dad6/raw/dad6b90975cea80460a244cfe03d78a7ca526a7b/log.txt)

File: Source/ASImageNode+AnimatedImage.mm [hunk 1]
```diff
@@ -53,6 +53,14 @@ - (void)_locked_setAnimatedImage:(id <ASAnimatedImageProtocol>)animatedImage
   if (ASObjectIsEqual(_animatedImage, animatedImage)) {
     return;
   }
+    
+  if (animatedImage == nil) {
```
Comment (nguyenhuy) [inline Source/ASImageNode+AnimatedImage.mm]: I don't understand this check. Shouldn't we do this whenever `_animatedImage != nil`?

File: Source/ASImageNode+AnimatedImage.mm [hunk 2]
```diff
@@ -53,6 +53,14 @@ - (void)_locked_setAnimatedImage:(id <ASAnimatedImageProtocol>)animatedImage
...
+    // Animated image can take while to dealloc, do it off the main queue
+    __block id <ASAnimatedImageProtocol>deallocImage = _animatedImage;
+    ASPerformBlockOnBackgroundThread(^{
```
Comment (nguyenhuy) [inline Source/ASImageNode+AnimatedImage.mm]: For consistency, let's use `ASPerformBackgroundDeallocation()`. Also, let's use `previousAnimatedImage` declared a few lines below.

Comment (nguyenhuy) [reply Source/ASImageNode+AnimatedImage.mm]: Lastly, I think it's easier to reason if we do this at the end of this method.

File: Source/ASNetworkImageNode.mm [hunk 3]
```diff
@@ -696,50 +696,61 @@ - (void)_lazilyLoadImageIfNecessary
...
-        as_log_verbose(ASImageLoadingLog(), "Downloaded image for %@ img: %@ urls: %@", self, [imageContainer asdk_image], URLs);
-        
-        // Grab the lock for the rest of the block
-        ASDN::MutexLocker l(strongSelf->__instanceLock__);
-        
-        //Getting a result back for a different download identifier, download must not have been successfully canceled
-        if (ASObjectIsEqual(strongSelf->_downloadIdentifier, downloadIdentifier) == NO && downloadIdentifier != nil) {
-          return;
-        }
+        ASPerformBlockOnBackgroundThread(^{
```
Comment (nguyenhuy) [inline Source/ASNetworkImageNode.mm]: Just to confirm, the only change here is that the (previous) `finished` block is now executed on a background thread? (it's a bit hard to read)

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Any source code changes should have an entry in CHANGELOG.md or have #trivial in their title.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: Source/ASImageNode+AnimatedImage.mm (hunk 1 above)
Comment (garrettmoon) [reply Source/ASImageNode+AnimatedImage.mm]: Yeeeeees, good catch.

File: Source/ASNetworkImageNode.mm (hunk 3 above)
Comment (garrettmoon) [reply Source/ASNetworkImageNode.mm]: Mostly. Everything before was happening on main. Now it's being dispatched to the background and once it's done the work, the delegate callbacks are dispatched to main.

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Any source code changes should have an entry in CHANGELOG.md or have #trivial in their title.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by nguyenhuy** [comment]
@garrettmoon There are conflicts that I'm not confident in fixing. Please resolve them and merge away.

File: Source/ASNetworkImageNode.mm [hunk 4]
```diff
@@ -696,50 +696,61 @@ - (void)_lazilyLoadImageIfNecessary
...
+            return;
           }
-        } else if (error && strongSelf->_delegateFlags.delegateDidFailWithError) {
-          ASDN::MutexUnlocker u(strongSelf->__instanceLock__);
-          [delegate imageNode:strongSelf didFailWithError:error];
-        }
+          
+          if (imageContainer != nil) {
+            [strongSelf _locked_setCurrentImageQuality:1.0];
+            if ([imageContainer asdk_animatedImageData] && strongSelf->_downloaderFlags.downloaderImplementsAnimatedImage) {
```
Comment (maicki) [inline Source/ASNetworkImageNode.mm]: You could move up calling `asdk_animatedImageData` and assign it to a variable so you don't have to call it twice if this evaluates to true. Don't know how heavy `asdk_animatedImageData` is though.

**Comment by nguyenhuy** [comment]
@garrettmoon  Ping :)

**Comment by garrettmoon** [comment]
@nguyenhuy Ah! Will resolve and land today, thank you all for the review!

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Any source code changes should have an entry in CHANGELOG.md or have #trivial in their title.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by pinterest-ios** [comment]
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/de9d5c5807d1274629aa43aac12da8ab/raw/4ea14d1689e1bf41dcc4265af00b7572080767c0/log.txt)

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Any source code changes should have an entry in CHANGELOG.md or have #trivial in their title.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: Source/ASImageNode+AnimatedImage.mm [hunk 5]
```diff
@@ -79,6 +80,11 @@ - (void)_locked_setAnimatedImage:(id <ASAnimatedImageProtocol>)animatedImage
       [self setCoverImage:nil];
   }
   
+  // Animated image can take while to dealloc, do it off the main queue
+  if (previousAnimatedImage != nil) {
+    ASPerformBackgroundDeallocation(&previousAnimatedImage);
+  }
+  
   [self animatedImageSet:_animatedImage previousAnimatedImage:previousAnimatedImage];
```
Comment (Adlai-Holler) [inline Source/ASImageNode+AnimatedImage.mm]: `previousAnimatedImage` will be nilled out by this function. Call this after calling `-animatedImageSet:previousAnimatedImage:`

Comment (garrettmoon) [reply Source/ASImageNode+AnimatedImage.mm]: Oooh, good catch!

**Pull request merged**